### ISSUE TITLE
Better request matching

### ISSUE THREAD

**Issue by gaborcsardi** [gaborcsardi/httrmock#3]
Maybe. But maybe we can just keep this package simple, so not until I really need this.

**Comment by gaborcsardi** [comment]
We should probably only match on the URL and the HTTP method, at least by default...

**Comment by gaborcsardi** [comment]
59c64a5d0883134385dd53191eaea71ae3a41042 only matches on the URL and the method.

**Comment by jennybc** [comment]
> so not until I really need this

I fear that day has come.

I have a problem testing `gh_whoami()`. Each call spawns a GET to `https://api.github.com/user`. So the first call makes a recording. But then that gets replayed for all subsequent calls, even though I am changing the token and expecting different outcomes. 😢

**Comment by jennybc** [comment]
Right now httrmock doesn't know anything about testthat. But if it's not too odious, maybe the `desc` of `test_that()` could be added to the key? I can understand why you don't want to use the entire request. It seems reasonable that calls to same URL with same HTTP method should live in different tests.

**Comment by gaborcsardi** [comment]
Hmmm. In general the token is not part of the matching, because when the tests are running in the wild, there are no tokens. 

I suggest that as a workaround, you could simply pass an extra header to the calls, with a different value each time.

**Comment by jennybc** [comment]
But the headers aren't consulted in the matching. Only `url` and `method`.

``` r
request_key <- function(req) {
  digest(list(method = req$method, url = req$url))
}
```

I don't see how this will help. Am I confused?

**Comment by gaborcsardi** [comment]
Right, sorry I forgot that.

**Comment by gaborcsardi** [comment]
Originally they were part of the matching, but I changed it later.

**Comment by jennybc** [comment]
It's hard to know what to put in there. 🤔

**Comment by gaborcsardi** [comment]
So the problem with including all headers is that they often differ on various systems. E.g. user-agent contains the httr version, and the curl library version as well AFAIR. 

But clearly, we need a mechanism to put some "stuff" in the matching. Either a mechanism to whitelist some headers and include them in the hash, or another way or adding arbitrary context.

**Comment by jennybc** [comment]
That's why I was wondering if `desc` could be sought in the parent environment (or wherever it would show up). I see that is a bit icky though.

**Comment by gaborcsardi** [comment]
Yeah, I would rather have a generic mechanism, that works well without testthat as well.

**Comment by gaborcsardi** [comment]
So how about a `with_httmock_context` function, that evaluates an arbitrary R expression in a specific `httrmock` context, and the context is part of the matching?

**Comment by jennybc** [comment]
That works.

I wonder if there's a different word to head off any confusion with testthat context?

Would or could these contexts (or whatever) have associated slugs? I think that might be a nice addition to the current cryptic keys and object names. This sort of matters when you're deleting, staging, and committing recordings.

**Comment by gaborcsardi** [comment]
> I wonder if there's a different word to head off any confusion with testthat context?

If you have a better idea, let me know.  I think with the `with_httrmock_` prefix it is OK.

> Would or could these contexts (or whatever) have associated slugs? I think that might be a nice addition to the current cryptic keys and object names. This sort of matters when you're deleting, staging, and committing recordings.

Yeah, unfortunately the `storr` backend is not ideal for this. We would probably need a slightly different backend, I was meaning to write Rich about this. And yes, great idea, the context slugs should be definitely part of the file name. 

Actually, we don't even need the full power of `storr`, just a file-based key-value data store, really. But `storr` was very convenient to use and saved me a lot of time.

**Comment by gaborcsardi** [comment]
As for better request matching, we could define what to do with different headers, and parts of the request in general:
* drop it
* use it in a key (i.e. in the matching)
* transform it and then use it in a key
* don't use it in a key, but keep it
* transform it and keep it (but don't use it in a key)

This would be user configurable, and we would provide a sensible default that drops `Authorization`, etc.
